File: backend/app/schemas/game.py
import uuid
from datetime import date, datetime
from pydantic import BaseModel
from app.models.game import SwipeAction
from app.schemas.card import CardOut


class GameSessionOut(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    user_id: uuid.UUID
    persona_id: uuid.UUID | None = None
    stage: int
    progress: float
    topic_mastery: dict
    investor_rank: int
    capital: float
    portfolio_weights: dict
    peak_capital: float
    is_daily: bool
    daily_date: date | None = None
    daily_cards_played: int
    daily_target: int
    daily_completed: bool
    streak_bonus_awarded: float
    created_at: datetime
    updated_at: datetime


class GameEventOut(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    card_id: int | None
    action: str
    reward: float
    card: CardOut | None = None
    created_at: datetime


class SwipeRequest(BaseModel):
    card_id: int
    action: SwipeAction


class SwipeResponse(BaseModel):
    lesson: str
    reward: float
    session: GameSessionOut
    next_card: CardOut | None = None


class DailyStatusOut(BaseModel):
    streak_count: int
    cards_completed_today: int
    daily_target: int
    remaining_cards: int
    completed_today: bool
    streak_bonus_capital: float = 1000.0

File: frontend/services/game.ts
import { api } from "./api";

export interface CardData {
  id: number;
  type: "education" | "event" | "action";
  title: string;
  body: string;
  emoji: string;
  stage_min: number;
  stage_max: number;
  topics: string[];
  linked_traits: string[];
  difficulty: number;
  diagnostic_power: number;
  left_choice: string;
  right_choice: string;
  left_lesson: string;
  right_lesson: string;
  card_band_color: string;
}

export interface GameEventData {
  id: number;
  card_id: number | null;
  action: "left" | "right";
  reward: number;
  card: CardData | null;
  created_at: string;
}

export interface SessionData {
  id: string;
  user_id: string;
  persona_id: string | null;
  stage: number;
  progress: number;
  topic_mastery: Record<string, number>;
  investor_rank: number;
  capital: number;
  portfolio_weights: Record<string, number>;
  peak_capital: number;
  created_at: string;
  updated_at: string;
}

export interface SwipeResponse {
  lesson: string;
  reward: number;
  session: SessionData;
  next_card: CardData | null;
}

export async function getSessions(): Promise<SessionData[]> {
  const resp = await api.get<SessionData[]>("/api/game/sessions");
  return resp.data;
}

export async function createSession(): Promise<SessionData> {
  const resp = await api.post<SessionData>("/api/game/sessions");
  return resp.data;
}

export async function getSession(sessionId: string): Promise<SessionData> {
  const resp = await api.get<SessionData>(`/api/game/sessions/${sessionId}`);
  return resp.data;
}

export async function swipe(
  sessionId: string,
  cardId: number,
  action: "left" | "right"
): Promise<SwipeResponse> {
  const resp = await api.post<SwipeResponse>(
    `/api/game/sessions/${sessionId}/swipe`,
    { card_id: cardId, action }
  );
  return resp.data;
}

export async function getNextCard(sessionId: string): Promise<CardData | null> {
  const resp = await api.get<CardData | null>(`/api/game/sessions/${sessionId}/next-card`);
  return resp.data;
}

export async function getSessionHistory(sessionId: string): Promise<GameEventData[]> {
  const resp = await api.get<GameEventData[]>(`/api/game/sessions/${sessionId}/history`);
  return resp.data;
}
